Retry fails with NoAvailableWorkerLeaseException

## Reproducer

Checkout https://github.com/leonard84/spock/tree/retry-fail and run `gw :sSpec:test --stacktrace --tests TimeoutExtension`

## Buildscan
https://ge.spockframework.org/s/myomimtysi5cy/failure#1

## Exception
```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':spock-specs:test'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:147)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:145)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:133)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:333)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:320)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:313)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:299)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:143)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:227)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:218)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:140)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
Caused by: org.gradle.internal.work.NoAvailableWorkerLeaseException: No worker lease associated with the current thread
        at org.gradle.internal.work.DefaultWorkerLeaseService.getCurrentWorkerLease(DefaultWorkerLeaseService.java:91)
        at org.gradle.internal.operations.DefaultBuildOperationQueueFactory.create(DefaultBuildOperationQueueFactory.java:33)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.executeInParallel(DefaultBuildOperationExecutor.java:132)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.runAll(DefaultBuildOperationExecutor.java:102)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ParallelResolveArtifactSet$VisitingSet.visit(ParallelResolveArtifactSet.java:66)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.visitArtifacts(DefaultLenientConfiguration.java:296)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.access$500(DefaultLenientConfiguration.java:74)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$3.run(DefaultLenientConfiguration.java:244)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.visitArtifactsWithBuildOperation(DefaultLenientConfiguration.java:241)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.access$200(DefaultLenientConfiguration.java:74)
        at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$2.visitArtifacts(DefaultLenientConfiguration.java:144)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.visitContents(DefaultConfiguration.java:1473)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.visitContents(DefaultConfiguration.java:509)
        at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:351)
        at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:119)
        at org.gradle.api.internal.file.collections.UnpackingVisitor.add(UnpackingVisitor.java:64)
        at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection$UnresolvedItemsCollector.visitContents(DefaultConfigurableFileCollection.java:372)
        at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.visitChildren(DefaultConfigurableFileCollection.java:284)
        at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:119)
        at org.gradle.api.internal.file.AbstractFileCollection.getFiles(AbstractFileCollection.java:127)
        at org.gradle.api.internal.file.AbstractFileCollection.iterator(AbstractFileCollection.java:173)
        at org.gradle.testretry.internal.testsreader.TestsReader.readClasspathClass(TestsReader.java:80)
        at org.gradle.testretry.internal.testsreader.TestsReader.readClass(TestsReader.java:59)
        at org.gradle.testretry.internal.filter.AnnotationInspectorImpl.getClassAnnotations(AnnotationInspectorImpl.java:46)
        at org.gradle.testretry.internal.filter.AnnotationInspectorImpl$ClassAnnotationVisitor.visit(AnnotationInspectorImpl.java:78)
        at org.gradle.testretry.org.objectweb.asm.ClassReader.accept(ClassReader.java:569)
        at org.gradle.testretry.org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
        at org.gradle.testretry.internal.testsreader.TestsReader.visit(TestsReader.java:73)
        at org.gradle.testretry.internal.testsreader.TestsReader.visitClassFile(TestsReader.java:65)
        at org.gradle.testretry.internal.testsreader.TestsReader.lambda$readTestClassDirClass$1(TestsReader.java:51)
        at org.gradle.testretry.internal.testsreader.TestsReader.readTestClassDirClass(TestsReader.java:51)
        at org.gradle.testretry.internal.testsreader.TestsReader.readClass(TestsReader.java:55)
        at org.gradle.testretry.internal.filter.AnnotationInspectorImpl.getClassAnnotations(AnnotationInspectorImpl.java:46)
        at org.gradle.testretry.internal.filter.AnnotationInspectorImpl$ClassAnnotationVisitor.visit(AnnotationInspectorImpl.java:78)
        at org.gradle.testretry.org.objectweb.asm.ClassReader.accept(ClassReader.java:569)
        at org.gradle.testretry.org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
        at org.gradle.testretry.internal.testsreader.TestsReader.visit(TestsReader.java:73)
        at org.gradle.testretry.internal.testsreader.TestsReader.visitClassFile(TestsReader.java:65)
        at org.gradle.testretry.internal.testsreader.TestsReader.lambda$readTestClassDirClass$1(TestsReader.java:51)
        at org.gradle.testretry.internal.testsreader.TestsReader.readTestClassDirClass(TestsReader.java:51)
        at org.gradle.testretry.internal.testsreader.TestsReader.readClass(TestsReader.java:55)
        at org.gradle.testretry.internal.filter.AnnotationInspectorImpl.getClassAnnotations(AnnotationInspectorImpl.java:46)
        at org.gradle.testretry.internal.filter.RetryFilter.canRetry(RetryFilter.java:59)
        at org.gradle.testretry.internal.executer.RetryTestResultProcessor.failure(RetryTestResultProcessor.java:133)
        at org.gradle.api.internal.tasks.testing.results.AttachParentTestResultProcessor.failure(AttachParentTestResultProcessor.java:43)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.FailureHandlingDispatch.dispatch(FailureHandlingDispatch.java:30)
        at org.gradle.internal.dispatch.AsyncDispatch.dispatchMessages(AsyncDispatch.java:87)
        at org.gradle.internal.dispatch.AsyncDispatch.access$000(AsyncDispatch.java:36)
        at org.gradle.internal.dispatch.AsyncDispatch$1.run(AsyncDispatch.java:71)
        at org.gradle.internal.concurrent.InterruptibleRunnable.run(InterruptibleRunnable.java:42)
        at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)


```

--- a/plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java
+++ b/plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java
@@ -100,7 +100,7 @@ public final class TestTaskConfigurer {
     private static RetryTestExecuter createRetryTestExecuter(Test task, TestRetryTaskExtensionAdapter extension, ObjectFactory objectFactory) {
         TestExecuter<JvmTestExecutionSpec> delegate = getTestExecuter(task);
         Instantiator instantiator = invoke(declaredMethod(AbstractTestTask.class, "getInstantiator"), task);
-        return new RetryTestExecuter(task, extension, delegate, instantiator, objectFactory);
+        return new RetryTestExecuter(task, extension, delegate, instantiator, objectFactory, task.getTestClassesDirs().getFiles(), task.getClasspath().getFiles());
     }
 
     private static TestExecuter<JvmTestExecutionSpec> getTestExecuter(Test task) {

--- a/plugin/src/main/java/org/gradle/testretry/internal/executer/RetryTestExecuter.java
+++ b/plugin/src/main/java/org/gradle/testretry/internal/executer/RetryTestExecuter.java
@@ -27,6 +27,8 @@ import org.gradle.testretry.internal.executer.framework.TestFrameworkStrategy;
 import org.gradle.testretry.internal.filter.AnnotationInspectorImpl;
 import org.gradle.testretry.internal.filter.RetryFilter;
 
+import java.io.File;
+import java.util.Set;
 import java.util.stream.Collectors;
 
 import static org.gradle.testretry.internal.executer.framework.TestFrameworkStrategy.gradleVersionIsAtLeast;
@@ -45,7 +47,9 @@ public final class RetryTestExecuter implements TestExecuter<JvmTestExecutionSpe
         TestRetryTaskExtensionAdapter extension,
         TestExecuter<JvmTestExecutionSpec> delegate,
         Instantiator instantiator,
-        ObjectFactory objectFactory
+        ObjectFactory objectFactory,
+        Set<File> testClassesDir,
+        Set<File> resolvedClasspath
     ) {
         this.extension = extension;
         this.delegate = delegate;
@@ -53,7 +57,9 @@ public final class RetryTestExecuter implements TestExecuter<JvmTestExecutionSpe
         this.frameworkTemplate = new TestFrameworkTemplate(
             testTask,
             instantiator,
-            objectFactory
+            objectFactory,
+            testClassesDir,
+            resolvedClasspath
         );
     }
 

--- a/plugin/src/main/java/org/gradle/testretry/internal/executer/TestFrameworkTemplate.java
+++ b/plugin/src/main/java/org/gradle/testretry/internal/executer/TestFrameworkTemplate.java
@@ -20,6 +20,9 @@ import org.gradle.api.tasks.testing.Test;
 import org.gradle.internal.reflect.Instantiator;
 import org.gradle.testretry.internal.testsreader.TestsReader;
 
+import java.io.File;
+import java.util.Set;
+
 public class TestFrameworkTemplate {
 
     public final Test task;
@@ -27,11 +30,11 @@ public class TestFrameworkTemplate {
     public final ObjectFactory objectFactory;
     public final TestsReader testsReader;
 
-    public TestFrameworkTemplate(Test task, Instantiator instantiator, ObjectFactory objectFactory) {
+    public TestFrameworkTemplate(Test task, Instantiator instantiator, ObjectFactory objectFactory, Set<File> testClassesDir, Set<File> resolvedClasspath) {
         this.task = task;
         this.instantiator = instantiator;
         this.objectFactory = objectFactory;
-        this.testsReader = new TestsReader(task.getTestClassesDirs().getFiles(), task.getClasspath());
+        this.testsReader = new TestsReader(testClassesDir, resolvedClasspath);
     }
 
     public TestFilterBuilder filterBuilder() {

--- a/plugin/src/test/groovy/org/gradle/testretry/FilterFuncTest.groovy
+++ b/plugin/src/test/groovy/org/gradle/testretry/FilterFuncTest.groovy
@@ -97,10 +97,11 @@ class FilterFuncTest extends AbstractGeneralPluginFuncTest {
             import java.lang.annotation.Retention;
             import java.lang.annotation.RetentionPolicy;
             import java.lang.annotation.Target;
+            import java.lang.annotation.Inherited;
 
             @Retention(RetentionPolicy.RUNTIME)
             @Target({ ElementType.TYPE })
-            @java.lang.annotation.Inherited
+            @Inherited
             public @interface InheritedExcludedAnnotation { }
         """
 
@@ -148,10 +149,11 @@ class FilterFuncTest extends AbstractGeneralPluginFuncTest {
             import java.lang.annotation.Retention;
             import java.lang.annotation.RetentionPolicy;
             import java.lang.annotation.Target;
+            import java.lang.annotation.Inherited;
 
             @Retention(RetentionPolicy.RUNTIME)
             @Target({ ElementType.TYPE })
-            @java.lang.annotation.Inherited
+            @Inherited
             public @interface $name { }
         """
     }

--- a/plugin/src/test/groovy/org/gradle/testretry/FilterFuncTest.groovy
+++ b/plugin/src/test/groovy/org/gradle/testretry/FilterFuncTest.groovy
@@ -67,7 +67,7 @@ class FilterFuncTest extends AbstractGeneralPluginFuncTest {
         gradleVersion << GRADLE_VERSIONS_UNDER_TEST
     }
 
-    def "annotation can be inherited from lasspath"() {
+    def "annotation can be inherited from classpath"() {
         given:
         settingsFile << """
             include "lib"

--- a/plugin/src/test/groovy/org/gradle/testretry/FilterFuncTest.groovy
+++ b/plugin/src/test/groovy/org/gradle/testretry/FilterFuncTest.groovy
@@ -67,6 +67,66 @@ class FilterFuncTest extends AbstractGeneralPluginFuncTest {
         gradleVersion << GRADLE_VERSIONS_UNDER_TEST
     }
 
+    def "annotation can be inherited from lasspath"() {
+        given:
+        settingsFile << """
+            include "lib"
+        """
+
+        file("lib/build.gradle") << """
+            plugins { id "java" }
+        """
+
+        buildFile << """
+            dependencies {
+                testImplementation(project(":lib"))
+            }
+            
+            test.retry {
+                maxRetries = 2
+                filter {
+                    excludeAnnotationClasses.add("*Excluded*")
+                }
+            }
+        """
+
+        and:
+        file("lib/src/main/java/InheritedExcludedAnnotation.java") << """
+            package acme;
+            import java.lang.annotation.ElementType;
+            import java.lang.annotation.Retention;
+            import java.lang.annotation.RetentionPolicy;
+            import java.lang.annotation.Target;
+
+            @Retention(RetentionPolicy.RUNTIME)
+            @Target({ ElementType.TYPE })
+            @java.lang.annotation.Inherited
+            public @interface InheritedExcludedAnnotation { }
+        """
+
+        file("lib/src/main/java/acme/BaseTest.java") << """
+            package acme;
+            @InheritedExcludedAnnotation
+            public class BaseTest {
+            
+            }
+        """
+
+        and:
+        def noRetry = []
+        noRetry << test("ExcludedTest", "BaseTest",)
+
+        when:
+        def result = gradleRunner(gradleVersion).buildAndFail()
+
+        then:
+        result.output.count("acme.ExcludedTest > flakyTest FAILED") == 1
+        result.output.count("acme.ExcludedTest > flakyTest PASSED") == 0
+
+        where:
+        gradleVersion << GRADLE_VERSIONS_UNDER_TEST
+    }
+
     private void nonInheritedAnnotation(String name) {
         file("src/test/java/acme/${name}.java") << """
             package acme;
@@ -81,8 +141,8 @@ class FilterFuncTest extends AbstractGeneralPluginFuncTest {
         """
     }
 
-    private void inheritedAnnotation(String name) {
-        file("src/test/java/${name}.java") << """
+    private void inheritedAnnotation(String name, String project = "") {
+        file("${project ? project + "/" : ""}src/test/java/${name}.java") << """
             package acme;
             import java.lang.annotation.ElementType;
             import java.lang.annotation.Retention;

